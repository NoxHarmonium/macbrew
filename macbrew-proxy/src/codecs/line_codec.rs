use std::{io, str};
use tokio_util::codec::{Decoder, Encoder};

use bytes::BytesMut;

pub struct LineCodec;

impl Decoder for LineCodec {
    type Item = String;
    type Error = io::Error;

    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let newline = src.as_ref().iter().position(|b| *b == b'\n');
        if let Some(n) = newline {
            let line = src.split_to(n + 1);
            return match str::from_utf8(line.as_ref()) {
                Ok(s) => Ok(Some(s.to_string())),
                Err(_) => Err(io::Error::new(io::ErrorKind::Other, "Invalid String")),
            };
        }
        Ok(None)
    }
}

impl Encoder<String> for LineCodec {
    type Error = io::Error;

    fn encode(&mut self, item: String, dst: &mut BytesMut) -> Result<(), Self::Error> {
        Ok(())
    }
}
