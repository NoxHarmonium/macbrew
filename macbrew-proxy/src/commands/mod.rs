pub mod command;
pub mod get_fermentation;
pub mod get_recipes;
pub mod get_sessions;
pub mod list_sessions;
pub mod list_steps;
pub mod ping;
