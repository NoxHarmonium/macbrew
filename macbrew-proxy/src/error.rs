use reqwest::Error as ReqwestError;
use serde::{ser, Serialize};
use snafu::Snafu;

// TODO: Clean these up to provide more context and better error messages
#[derive(Debug, Snafu, Serialize)]
pub enum Error {
    #[snafu(display("Could not serialize data"))]
    SerializationFailure {},

    #[snafu(display("Error calling API: {}", inner_error))]
    ApiError {
        #[serde(skip_serializing)]
        inner_error: ReqwestError,
    },

    #[snafu(display("API response failed validation: {}", message))]
    ApiResponseValidationError { message: String },

    #[snafu(display("Invalid command input: {}", message))]
    InvalidCommandInput { message: String },

    #[snafu(display("Serial IO error: {}", inner_error))]
    SerialIoError {
        #[serde(skip_serializing)]
        inner_error: std::io::Error,
    },

    #[snafu(display("Unknown error: {}", message))]
    UnknownError { message: String },

    #[snafu(display("Could not allocate serialization buffer. (Not enough RAM?)"))]
    SerializeBufferFull {},

    #[snafu(display("Could not work out length of sequence"))]
    SerializeSeqLengthUnknown {},
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

impl From<ReqwestError> for Error {
    fn from(inner_error: ReqwestError) -> Self {
        Error::ApiError {
            inner_error: inner_error,
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(inner_error: std::io::Error) -> Self {
        Error::SerialIoError {
            inner_error: inner_error,
        }
    }
}

impl ser::Error for Error {
    #[cold]
    fn custom<T: std::fmt::Display>(msg: T) -> Error {
        // TODO: What should this actually be doing?
        // Hint: https://github.com/serde-rs/json/blob/master/src/error.rs#L373
        Error::UnknownError {
            message: msg.to_string(),
        }
    }
}
