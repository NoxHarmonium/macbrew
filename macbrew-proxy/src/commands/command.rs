use crate::data::data_manager::DataManager;
use crate::error::Result;
use crate::serializers::mac::serialize;
use async_trait::async_trait;
use serde::Serialize;

#[async_trait]
pub trait Command<DataManagerType: DataManager> {
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>>;
}

pub fn prepare_response<T>(rid: &str, success: bool, response_object: &T) -> Result<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    let preable_bytes = serialize(&rid)?;
    let success_bytes = serialize(&success)?;
    let response_bytes = serialize(&response_object)?;
    Ok([preable_bytes, success_bytes, response_bytes].concat())
}
