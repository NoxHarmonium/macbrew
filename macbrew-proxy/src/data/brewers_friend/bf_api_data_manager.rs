use crate::data::brewers_friend::bf_data_manager::BFDataManager;
use crate::data::brewers_friend::recipes::BeerXml;
use crate::data::brewers_friend::sessions::{BFBrewSession, BFBrewSessionsResponse};
use crate::error::{Error, Result};
use async_trait::async_trait;
use reqwest;
use reqwest::header::HeaderMap;
use serde::Deserialize;

#[cfg(test)]
use mockito;

#[derive(Deserialize, Debug)]
struct Config {
    brewers_friend_api_key: String,
}

pub struct BFApiDataManager;

fn get_base_url() -> String {
    #[cfg(test)]
    return mockito::server_url();
    #[cfg(not(test))]
    return String::from("https://api.brewersfriend.com");
}

fn create_client() -> Result<reqwest::Client> {
    let mut headers = HeaderMap::new();

    let config = match envy::from_env::<Config>() {
        Ok(config) => config,
        Err(error) => panic!("Missing environment variables: {:#?}", error),
    };

    headers.insert("X-API-KEY", config.brewers_friend_api_key.parse().unwrap());

    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    Ok(client)
}

#[async_trait]
impl BFDataManager for BFApiDataManager {
    async fn sessions() -> Result<Vec<BFBrewSession>> {
        println!("Fetching sessions...");
        let client = create_client()?;

        let brew_sessions_url = format!("{}/v1/brewsessions", get_base_url());

        let res = client.get(&brew_sessions_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched [{:#?}] sessions", response.brewsessions.len());

        Ok(response.brewsessions)
    }

    async fn session(id: &str) -> Result<BFBrewSession> {
        println!("Fetching session with id [{}]...", id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/brewsessions/{}", get_base_url(), id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        let session = match response.brewsessions.as_slice() {
            [] => Err(Error::ApiResponseValidationError {
                message: String::from("No session with ID found"),
            }),
            [session] => Ok(session.to_owned()),
            [_args @ ..] => Err(Error::ApiResponseValidationError {
                message: String::from("Multiple sessions for ID found. This should not happen"),
            }),
        };
        println!("Fetched session with id [{}]...", id);

        session
    }

    async fn recipe(recipe_id: &str) -> Result<BeerXml> {
        println!("Fetching recipe with id [{}]...", recipe_id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/recipes/{}.xml", get_base_url(), recipe_id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response = match quick_xml::de::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched recipe with id [{}]...", recipe_id);

        Ok(response)
    }
}
