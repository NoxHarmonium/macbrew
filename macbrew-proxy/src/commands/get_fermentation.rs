use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BfDataManager;
use crate::data::macbrew::fermentation::FermentationData;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetFermentationDataCommand<DataManager: BfDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BfDataManager + 'static> Command<DataManager>
    for GetFermentationDataCommand<DataManager>
{
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [session_id] => {
                let bf_fermentation = DataManager::fermentation(session_id).await?;
                let fermentation_data = FermentationData::from_bf_fermentation(&bf_fermentation);
                prepare_response(rid, true, &fermentation_data)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Session ID)"),
            }),
        }
    }
}
