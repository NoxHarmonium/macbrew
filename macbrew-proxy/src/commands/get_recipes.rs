use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BfDataManager;
use crate::data::macbrew::recipes::Recipe;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetRecipesCommand<DataManager: BfDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BfDataManager + 'static> Command<DataManager> for GetRecipesCommand<DataManager> {
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [recipe_id] => {
                let beer_xml_recipe = DataManager::recipe(recipe_id).await?;
                let recipe = Recipe::from_beer_xml(&beer_xml_recipe);
                prepare_response(rid, true, &recipe)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Recipe ID)"),
            }),
        }
    }
}
