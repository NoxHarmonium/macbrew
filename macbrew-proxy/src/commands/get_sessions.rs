use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BFDataManager;
use crate::data::macbrew::sessions::BrewSession;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetSessionsCommand<DataManager: BFDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BFDataManager + 'static> Command<DataManager>
    for GetSessionsCommand<DataManager>
{
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [session_id] => {
                let bf_session = DataManager::session(session_id).await?;
                let session = BrewSession::from_bf_session(&bf_session);
                prepare_response(rid, true, &session)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Session ID)"),
            }),
        }
    }
}
