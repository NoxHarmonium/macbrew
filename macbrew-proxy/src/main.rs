#![warn(clippy::all)]

use clap::Clap;
use error::{Error, Result};
use futures::sink::SinkExt;
use std::str;
use tokio::stream::StreamExt;
use tokio_serial::{Serial, SerialPortSettings};
use tokio_util::codec::Framed;

mod codecs;
mod commands;
mod data;
mod error;
mod serializers;

#[cfg(unix)]
const DEFAULT_TTY: &str = "/dev/pts/1";
#[cfg(windows)]
const DEFAULT_TTY: &str = "COM1";

use codecs::line_codec::LineCodec;
use commands::command::{prepare_response, Command};
use data::brewers_friend::bf_api_data_manager::BFApiDataManager;

#[derive(Clap)]
#[clap(version = "0.1", author = "Sean Dawson <contact@seandawson.info>")]
struct Opts {
    #[clap(default_value = DEFAULT_TTY)]
    serial_device: String,
}

async fn handle_line(line: &str) -> Result<Vec<u8>> {
    let components = line.split(' ').collect::<Vec<_>>();
    let response = match components.as_slice() {
        [] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
        [rid, "LIST", "SESSION", args @ ..] => {
            commands::list_sessions::ListSessionsCommand::<BFApiDataManager>::handle(rid, args)
                .await
        }
        [rid, "GET", "SESSION", args @ ..] => {
            commands::get_sessions::GetSessionsCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [rid, "GET", "RECIPE", args @ ..] => {
            commands::get_recipes::GetRecipesCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [_, _args @ ..] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
    };
    match response {
        Ok(response) => Ok(response),
        Err(error) => prepare_response("ERROR", false, &error.to_string()),
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let opts: Opts = Opts::parse();

    println!("Starting MacBrew on port {}", opts.serial_device);

    let settings = SerialPortSettings::default();
    let mut port = Serial::from_path(opts.serial_device, &settings).unwrap();

    #[cfg(unix)]
    port.set_exclusive(false)
        .expect("Unable to set serial port exclusive to false");

    let mut framed = Framed::new(port, LineCodec);

    while let Some(line_result) = framed.next().await {
        let line = line_result.expect("Failed to read line");
        let response = handle_line(line.trim()).await?;
        framed.send(response).await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {

    use encoding::all::MAC_ROMAN;
    use encoding::{EncoderTrap, Encoding};
    use hexplay::HexViewBuilder;
    use mockito::mock;

    use super::*;

    fn assert_request_id(binary: &Vec<u8>, request_id: &'static str) -> () {
        assert_eq!(
            &binary[4..6],
            MAC_ROMAN
                .encode(request_id, EncoderTrap::Replace)
                .unwrap()
                .as_slice()
        );
    }

    fn format_binary_for_snap(binary: &Vec<u8>) -> Vec<String> {
        let view = HexViewBuilder::new(binary.as_slice())
            .row_width(16)
            .finish();
        // Split on new lines to make the snapshots easier to read
        view.to_string().split('\n').map(String::from).collect()
    }

    #[tokio::test]
    async fn test_get_recipe() {
        let recipe_xml = include_str!("../resources/sample_recipe.xml");
        let _m = mock("GET", "/v1/recipes/123456.xml")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(recipe_xml)
            .create();

        let result = handle_line("11 GET RECIPE 123456").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "11");
    }

    #[tokio::test]
    async fn test_list_sessions() {
        let sessions_json = include_str!("../resources/sample_sessions.json");
        let _m = mock("GET", "/v1/brewsessions")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(sessions_json)
            .create();

        let result = handle_line("22 LIST SESSION").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "22");
    }

    #[tokio::test]
    async fn test_get_sessions() {
        let session_json = include_str!("../resources/sample_session.json");
        let _m = mock("GET", "/v1/brewsessions/363597")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(session_json)
            .create();

        let result = handle_line("33 GET SESSION 363597").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "33");
    }
}
