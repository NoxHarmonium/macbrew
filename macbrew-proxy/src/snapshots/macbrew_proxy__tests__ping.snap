---
source: macbrew-proxy/src/main.rs
expression: format_binary_for_snap(&result)
---
[
    "00000000  00 0B 00 01 31 01 00 05 50 6F 6E 67 21 6E 63 40  | .♂.☺1☺.♣Pong!nc@ |",
    "00000010  6E                                               | n                |",
]
